src/replace_weights_with_variables.jl

```diff
@@ -0,0 +1,84 @@
+# Copyright (c) 2024: Triad National Security, LLC
+# Copyright (c) 2024: Oscar Dowson and contributors
+#
+# Use of this source code is governed by a BSD-style license that can be found
+# in the LICENSE.md file.
+
+"""
+    replace_weights_with_variables(
+        model::JuMP.AbstractModel,
+        predictor::AbstractPredictor,
+    )
+
+Convert `predictor` with trained weights into a predictor in which the weights
+are JuMP decision variables.
+
+This function is useful when you wish to use constrained optimization to train
+small to moderate neural networks.
+
+## Example
+
+```jldoctest
+julia> using JuMP, Flux, MathOptAI
+
+julia> chain = Flux.Chain(Flux.Dense(2 => 3), Flux.softmax);
+
+julia> model = Model();
+
+julia> @variable(model, x[i in 1:2] == i);
+
+julia> predictor = MathOptAI.build_predictor(chain)
+Pipeline with layers:
+ * Affine(A, b) [input: 2, output: 3]
+ * SoftMax()
+
+julia> predictor = MathOptAI.replace_weights_with_variables(model, predictor)
```

Comment:
How's this as an API? One thing for InfiniteOpt: we kinda need the `x` here to know what type of decision variables to add, but it doesn't really fit. Thoughts?